src/Microsoft.AspNetCore.Authentication/Events/BaseContext.cs

```diff
@@ -9,41 +9,70 @@ namespace Microsoft.AspNetCore.Authentication
     /// <summary>
     /// Base class used by other context classes.
     /// </summary>
-    public abstract class BaseContext
+    public abstract class BaseContext<TOptions> where TOptions : AuthenticationSchemeOptions
     {
+        private AuthenticationProperties _properties;
+
         /// <summary>
         /// Constructor.
         /// </summary>
-        /// <param name="context">The request context.</param>
-        protected BaseContext(HttpContext context)
+        /// <param name="context">The context.</param>
+        /// <param name="scheme">The authentication scheme.</param>
+        /// <param name="options">The authentication options associated with the scheme.</param>
+        protected BaseContext(HttpContext context, AuthenticationScheme scheme, TOptions options)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
+            if (scheme == null)
+            {
+                throw new ArgumentNullException(nameof(scheme));
+            }
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
 
             HttpContext = context;
+            Scheme = scheme;
+            Options = options;
         }
 
         /// <summary>
+        /// The authentication scheme.
+        /// </summary>
+        public AuthenticationScheme Scheme { get; }
+
+        /// <summary>
+        /// Gets the authentication options associated with the scheme.
+        /// </summary>
+        public TOptions Options { get; }
+
+        /// <summary>
         /// The context.
         /// </summary>
         public HttpContext HttpContext { get; }
 
         /// <summary>
         /// The request.
         /// </summary>
-        public HttpRequest Request
-        {
-            get { return HttpContext.Request; }
-        }
+        public HttpRequest Request => HttpContext.Request;
 
         /// <summary>
         /// The response.
         /// </summary>
-        public HttpResponse Response
+        public HttpResponse Response => HttpContext.Response;
+
+        /// <summary>
+        /// Gets or sets the <see cref="AuthenticationProperties"/>.
+        /// </summary>
+        public virtual AuthenticationProperties Properties
         {
-            get { return HttpContext.Response; }
+            get => _properties ?? (_properties = new AuthenticationProperties());
+            set => _properties = value;
         }
+
+        public EventResultState State { get; protected set; }
```

Comment:
Rename/make RemoteAuthenticationResult => HandleRequestAuthenticationResult and be the only concept.  Enums are annoying since if we do want to add something like a Bypass, we can't do it until a major version, unlike results where we can add a new property whenever we want and use it immediately.  We want to only have a single way to do flow, not two (Result + enum)

The public API on the contexts are the same and there's no more enum, and internally we are consistent with only results.